### ISSUE TITLE
Numeric comparisons triggering error in errorprone

### ISSUE THREAD

**Issue by lehphyro** [immutables/immutables#570]
The following class generates errorprone errors:
```java
@Value.Immutable
abstract class AbstractTestClass {
    public abstract Optional<BigDecimal> getNumericValue();
}
```
The error message is:
```
error: [NumericEquality] Numeric comparison using reference equality instead of value equality
    if (this.numericValue == newValue) return this;
                               ^
    (see http://errorprone.info/bugpattern/NumericEquality)
  Did you mean 'if (Objects.equals(this.numericValue, newValue)) return this;' or 'if (this.numericValue.equals(newValue)) return this;'?
```

**Comment by Stephan202** [comment]
Since this code is part of a "wither" method, this is not a functional defect, but it'd be nice to avoid the warning regardless. Because [`NumericEquality`](http://errorprone.info/bugpattern/NumericEquality) has severity `ERROR` it cannot simply be suppressed using `-XepDisableWarningsInGeneratedCode`. A workaround is to explicitly downgrade this particular check to `WARN` severity, but that's not a burden that should be placed on all users of both Immutables and Error Prone.

I see two possible solutions:
1. Update the generated code to use `#equals` for subtypes of `java.lang.Number`. This is my preferred solution, but perhaps this is deemed too expensive. (Because in the common case the values are probably not equal.)
2. Add `@SuppressWarnings("NumericEquality")` to "wither" methods accepting an (optional) subtype of `java.lang.Number`.

@elucash, I'm willing to try and open a PR for either approach. Which do you prefer? (Or do you perhaps see a third option?)

**Comment by Stephan202** [comment]
Thinking about this some more, using `#equals` for `java.lang.Number` subtypes seems correct only if those subtypes are immutable. So that would require a whitelisting approach (`BigInteger`, `BigDecimal`, `UnsignedInteger`, `UnsignedLong`...), but not resolve the issue altogether.

So `@SuppressWarnings("NumericEquality")` seems the way to go. I see two potential implementations:
1. Simply add this annotation to the top of the class, where currently by default `@SuppressWarnings({"all"})` is generated. (Note that Error Prone does not accept "all". Maybe it should?)
2. Add custom `@SuppressWarnings` annotations only on the methods that trigger the error.

The first solution is decidedly simpler. WDYT?

**Comment by elucash** [comment]
@Stephan202 thank you for looking into it. Well, we can do either way. I don't think that we should care about mutability of `java.lang.Number`, because even if mutable `equals` should be sane for them (even if I do think that `equals` doesn't make sense for mutable objects). Having `@SuppressWarnings("NumericEquality")` is also neat, but probably it's better to limit the scope to the the specific method. If you can take care of PR that would be great.
Here's the hint:

```java
  //  Add following method to TypeInstrospectionBase
  // ValueAttribute inherits from it
  // this will be considered for the type or for element types of container types
  public boolean isNumericObjectType() {
    ensureTypeIntrospected();
    return extendedClassesNames.contains(Number.class.getName());
  }
```
then you can plase if statement in the template of wither method (optional case)

```
   [if v.optionalType]
 ....
[if v.numericObjectType]
```

or if we're handling this by using `equals`, the easiest way is to add this to `hasSimpleScalarElementType`

```java
  public boolean hasSimpleScalarElementType() {
...     
        || isJdkSpecializedOptional()
        || isNumericObjectType();// <-- added
  }

```

**Comment by Stephan202** [comment]
@elucash, I see you already took care of it, thanks! (Busy weekend, so I'm revisiting this ticket only now.)

**Comment by elucash** [comment]
@Stephan202 yeah, just applied the simplest of possible fixes, anyway, thank you for support!

**Comment by elucash** [comment]
released in 2.4.4. thank you!
